### PR TITLE
Routing: JComponentRouterRulesMenu to find right Itemid

### PR DESCRIPTION
**Pull request by Hackwar**
This rule implements finding the right Itemid as a generic rule for all component routers that extend from JComponentRouterView. It makes most of the parts of the *HelperRoute classes obsolete.
### How to test
- Apply this change
- Change the router of com_content by replacing

``` php
class ContentRouter extends JComponentRouterBase
{
```

with

``` php
class ContentRouter extends JComponentRouterView
{
    function __construct($app = null, $menu = null) {

        $categories = new JComponentRouterViewconfiguration('categories');
        $categories->setKey('id');
        $this->registerView($categories);
        $category = new JComponentRouterViewconfiguration('category');
        $category->setKey('id')->setParent($categories, 'id')->setNestable()->addLayout('blog');
        $this->registerView($category);
        $article = new JComponentRouterViewconfiguration('article');
        $article->setKey('id')->setParent($category, 'catid');
        $this->registerView($article);
        $this->registerView(new JComponentRouterViewconfiguration('archive'));
        $this->registerView(new JComponentRouterViewconfiguration('featured'));
        $this->registerView(new JComponentRouterViewconfiguration('form'));

        parent::__construct($app, $menu);
        $this->attachRule(new JComponentRouterRulesMenu($this));
    }

    public function getCategorySegment($id, $query)
    {
        $category = JCategories::getInstance($this->getName())->get($id);
        if ($category)
        {
            return array_reverse($category->getPath());
        }

        return array();
    }

    public function getCategoriesSegment($id, $query)
    {
        return $this->getCategorySegment($id, $query);
    }

    public function getArticleSegment($id, $query)
    {
        return array($id);
    }
```
- Comment line 62 and 111 from /components/com_content/helpers/route.php to disable the lookup of the Itemid there.
- See that the right menu items are still found for the items

This was updated on May 22nd 2015. All discussions prior to that point have either been integrated or are obsolete.

This was made possible through the generous donation of the people mentioned in the following link via an Indiegogo campaign: http://joomlager.de/crowdfunding/5-contributors

### CODE REVIEW

**Comment by Hackwar** [comment]
I've done further fixes and implemented unittests for this rule.

File: libraries/cms/component/router/rules/menu.php [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+							$this->lookup[$language][$view] = $item->id;
+						}
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Dummymethod to fullfill the interface requirements
```
Comment (wilsonge) [inline libraries/cms/component/router/rules/menu.php]: So much ewwwwww here....

Comment (Hackwar) [inline libraries/cms/component/router/rules/menu.php]: Do you mean the dummy methods? Or the implementation above?

**Comment by johanjanssens** [comment]
@Hackwar About the remove rule comments you made above. I understand, care should be taken though that if no rules are there the URL would still work. Otherwise things could go very wrong, without a working router Joomla would no longer work. 

A solution would be to ensure that if no rules are present we fallback to a none-SEF form. Is that already the case ? If no, might be good to include this scenario also in the unit tests.

**Comment by Hackwar** [comment]
@johanjanssens I specifically chose JComponentRouterViewconfiguration since I did not want to use JComponentRouterView for this. I can see us wanting to have a JComponentRouterView class for different purposes.

Regarding the setName/setViewName: Please see the latest commits to this PR. Those have since been removed.

**Comment by johanjanssens** [comment]
@Hackwar About setName/setViewName() I missed those changes. Looking much better indeed API wise.

About JComponentRouterViewconfiguration vs JComponentRouterView. I understand, still not 100% happy with JComponentRouterViewconfiguration, any alternatives ? JComponentRouterConfig maybe, or JComponentRouterContext ?

**Comment by Hackwar** [comment]
I've updated this PR to work with the changed JComponentRouterView class. Please test.

**Comment by dgt41** [comment]
@Hackwar what shall we look for the output urls? just B/C or something more/else?

**Comment by Hackwar** [comment]
simply B/C. This should not change the URLs.

**Comment by Hackwar** [comment]
I've combined the changes from this and all other routing related PRs into a new PR: #7615 Please review and comment in the new PR. I'm closing this one, so that we can focuse on the new PR.

**Pull request closed without merging**